### PR TITLE
[Clang] [C23] Implement N2653: u8 strings are char8_t[]

### LINKED ISSUE #97202: u8 prefixed strings have type char instead of char8_t (aka unsigned char)
example: https://godbolt.org/z/KzPz9qj6M

for reference:
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2418.pdf
https://www.open-std.org/JTC1/SC22/WG14/www/docs/n2653.htm

### PR DESCRIPTION
**Pull request by MitalAshok**
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2653.htm

Closes #97202

### CODE REVIEW

**Comment by MitalAshok** [comment]
CC @AaronBallman

Also the clang-format issues are intentional to fit the style of the surrounding lines

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (AaronBallman) [inline clang/lib/Frontend/InitPreprocessor.cpp:1347]: Would it make more sense for `Char8` to be set by default based on `C23`?

Comment (MitalAshok) [reply clang/lib/Frontend/InitPreprocessor.cpp:1347]: Looks like that also controls if `char8_t` is a keyword/builtin type, which it isn't in C23. (`LangOpts.Char8` is also false in C++20 with `-fno-char8_t`)

File: clang/lib/Headers/stdatomic.h
Comment (MitalAshok) [inline clang/lib/Headers/stdatomic.h:112]: Should the definition of `atomic_char8_t` also be gated behind C23/C++20?

Also in C++20, this should be `_Atomic(char8_t)`, but the next two are also wrong in C++11 (`_Atomic(char16_t)`/`_Atomic(char32_t)`)

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (AaronBallman) [reply clang/lib/Frontend/InitPreprocessor.cpp:1347]: Ah! Good call, thank you!

File: clang/lib/Headers/stdatomic.h
Comment (AaronBallman) [reply clang/lib/Headers/stdatomic.h:112]: Let's address that one in a separate PR, but yes `atomic_char8_t` should be gated behind C23 because the identifier isn't reserved until C23. For C++, it should be gated behind C++20.

As for the definition, technically in C they should be using `char8_t` and friends as well, but those are defined in `uchar.h` which we don't vend with the compiler (should we be vending that one in freestanding?). GCC seems to use `unsigned char` directly rather than use a transitive include, so I think that approach is fine for us to take as well.

Comment (MitalAshok) [reply clang/lib/Headers/stdatomic.h:112]: It seems like this header doesn't properly implement [P0943R6](https://wg21.link/P0943R6) in C++23 (Where `#define _Atomic(T) std::atomic<T>` and all these typedefs should just be `using std::atomic_*`.

So `::atomic_char8_t` should be `std::atomic<char8_t>`, not `std::atomic<unsigned char>` (nor `unsigned char _Atomic`), but that is outside the scope of this PR. I'll just make this match the surrounding typedefs for now.

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (MitalAshok) [inline clang/lib/Frontend/InitPreprocessor.cpp:1169]: This matches GCC behaviour: https://godbolt.org/z/6Eax3eqrd

But I'm not sure why it's defined in C++ at all

Comment (AaronBallman) [reply clang/lib/Frontend/InitPreprocessor.cpp:1169]: Hmmm that is surprising, because it expands to `unsigned char` and not `char8_t` there: https://godbolt.org/z/Y96bPbj8e and this seems to have changed between GCC 8 and GCC 9: https://godbolt.org/z/EaEPKr59E

It may be worth opening a GCC bug to see if this was intentional or not.

Comment (MitalAshok) [reply clang/lib/Frontend/InitPreprocessor.cpp:1169]: According to https://gcc.gnu.org/legacy-ml/gcc-patches/2018-11/msg00295.html `__CHAR8_TYPE__` is "the underlying type of `char8_t`" in C++. `char8_t` in C++ was added in GCC9 which explains the difference between GCC8 and 9

And this made me realise `-std=c++11 -fchar8_t` sets `LangOpts.Char8` in C++11, which I think this patch handles fine (See also: #97601).
Also `-std=c23 -fchar8_t` is a mode that exists. It makes `typeof(u8'x')` `char8_t` and `typeof(u8"")` `char8_t[1]`, even with this patch, which I guess is fine. Pending #55373 if `-std=c23 -fchar8_t` even makes sense.

Comment (cor3ntin) [reply clang/lib/Frontend/InitPreprocessor.cpp:1169]: I'm certainly concerned with the impact of that in C++. It should not be defined in C++ (ie, char9_t is not unsigned char in C++)

File: clang/test/C/C2x/n2653.c
Comment (AaronBallman) [inline clang/test/C/C2x/n2653.c:2]: You need to add `-ffreestanding` here otherwise your include below is using whatever it finds on the system rather than our header.

**Review by AaronBallman — COMMENTED** [review]
Precommit CI has found relevant failures:
```
_bk;t=1720277739321********************

_bk;t=1720277739321Failed Tests (1):

_bk;t=1720277739321  Clang :: C/C2x/n2653.c

_bk;t=1720277739321

_bk;t=1720277739321

_bk;t=1720277739321Testing Time: 261.44s
```

File: clang/test/C/C2x/n2653.c
Comment (MitalAshok) [reply clang/test/C/C2x/n2653.c]: That still seems to fail on the CI:

```
******************** TEST 'Clang :: C/C23/n2653.c' FAILED ********************
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/bin/clang -cc1 -internal-isystem /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/lib/clang/19/include -nostdsysteminc -ffreestanding -verify=c23 -std=c23 /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c
+ /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/bin/clang -cc1 -internal-isystem /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/build/lib/clang/19/include -nostdsysteminc -ffreestanding -verify=c23 -std=c23 /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c
error: 'c23-error' diagnostics seen but not expected: 
  File /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-vhr99-1/llvm-project/github-pull-requests/clang/test/C/C23/n2653.c Line 11: missing
1 error generated.

--

********************
```

So I've just removed the test for <stdatomic.h> (`-nostdinc` seems to disable the clang header too)

Comment (AaronBallman) [reply clang/test/C/C2x/n2653.c]: I don't think we should remove the test for stdatomic,h, that's an important property for us to test and may be papering over a bug here. The only situation we should be getting the system header is:
```
#if __STDC_HOSTED__ &&                                                         \
    __has_include_next(<stdatomic.h>) &&                                       \
    (!defined(_MSC_VER) || (defined(__cplusplus) && __cplusplus >= 202002L))
# include_next <stdatomic.h>
#else
...
so `-ffreestanding` really should only include our header. What do you get from `-E` on that machine? Are we actually including the system header anyway, or did we mess up our header somehow?
```

Comment (MitalAshok) [reply clang/test/C/C2x/n2653.c]: I think I ran the wrong tests then misread some `-E` output... Should be fixed now (`__STDC_VERSION__ > 202311L` -> `__STDC_VERSION__ >= 202311L`)

**Review by AaronBallman — APPROVED** [review]
LGTM, but please wait for @cor3ntin to give a final sign-off before landing.

**Review by cor3ntin — APPROVED** [review]
LGTM!

File: clang/lib/Frontend/InitPreprocessor.cpp
Comment (AaronBallman) [reply clang/lib/Frontend/InitPreprocessor.cpp]: FWIW, I think https://github.com/llvm/llvm-project/issues/55373 should be fixed so that the option is rejected in C modes, but that's somewhat orthogonal to this patch.

**Pull request merged**
